File: src/server/lib/redact.ts
const SENSITIVE_KEY_PATTERNS = [
  /key/i,
  /secret/i,
  /token/i,
  /password/i,
  /credential/i,
  /api_key/i,
  /auth/i,
  /private/i,
]

const SENSITIVE_VALUE_PATTERNS = [
  /^sk-[a-zA-Z0-9-_]{20,}/, // Anthropic / OpenAI keys
  /^sk-ant-[a-zA-Z0-9-_]+/, // Anthropic API keys
  /^ghp_[a-zA-Z0-9]{36}/, // GitHub personal access tokens
  /^gho_[a-zA-Z0-9]{36}/, // GitHub OAuth tokens
  /^github_pat_[a-zA-Z0-9_]+/, // GitHub fine-grained PATs
  /^Bearer\s+\S+/i, // Bearer tokens
  /^xox[bpsa]-[a-zA-Z0-9-]+/, // Slack tokens
  /^eyJ[a-zA-Z0-9_-]+\.eyJ/, // JWTs
  /^AKIA[A-Z0-9]{16}/, // AWS access keys
  /^npm_[a-zA-Z0-9]{36}/, // npm tokens
]

function isSensitiveKey(key: string): boolean {
  return SENSITIVE_KEY_PATTERNS.some((pattern) => pattern.test(key))
}

function isSensitiveValue(value: string): boolean {
  return SENSITIVE_VALUE_PATTERNS.some((pattern) => pattern.test(value))
}

function redactString(value: string): string {
  if (isSensitiveValue(value)) {
    return '[REDACTED]'
  }
  return value
}

export function redactSensitiveValues<T>(data: T): T {
  if (data === null || data === undefined) {
    return data
  }

  if (typeof data === 'string') {
    return redactString(data) as T
  }

  if (Array.isArray(data)) {
    return data.map((item) => redactSensitiveValues(item)) as T
  }

  if (typeof data === 'object') {
    const result: Record<string, unknown> = {}
    for (const [key, value] of Object.entries(data as Record<string, unknown>)) {
      if (key === 'env' && typeof value === 'object' && value !== null) {
        const redactedEnv: Record<string, string> = {}
        for (const envKey of Object.keys(value as Record<string, string>)) {
          redactedEnv[envKey] = '[REDACTED]'
        }
        result[key] = redactedEnv
      } else if (isSensitiveKey(key) && typeof value === 'string') {
        result[key] = '[REDACTED]'
      } else {
        result[key] = redactSensitiveValues(value)
      }
    }
    return result as T
  }

  return data
}

File: server/lib/redact.go
package lib

import "regexp"

// sensitiveKeyPatterns holds compiled regexes for keys whose values should be redacted.
var sensitiveKeyPatterns = []*regexp.Regexp{
	regexp.MustCompile(`(?i)key`),
	regexp.MustCompile(`(?i)secret`),
	regexp.MustCompile(`(?i)token`),
	regexp.MustCompile(`(?i)password`),
	regexp.MustCompile(`(?i)credential`),
	regexp.MustCompile(`(?i)api_key`),
	regexp.MustCompile(`(?i)auth`),
	regexp.MustCompile(`(?i)private`),
}

// sensitiveValuePatterns holds compiled regexes for string values that look like secrets.
var sensitiveValuePatterns = []*regexp.Regexp{
	regexp.MustCompile(`^sk-[a-zA-Z0-9\-_]{20,}`),          // Anthropic / OpenAI keys
	regexp.MustCompile(`^sk-ant-[a-zA-Z0-9\-_]+`),          // Anthropic API keys
	regexp.MustCompile(`^ghp_[a-zA-Z0-9]{36}`),              // GitHub personal access tokens
	regexp.MustCompile(`^gho_[a-zA-Z0-9]{36}`),              // GitHub OAuth tokens
	regexp.MustCompile(`^github_pat_[a-zA-Z0-9_]+`),         // GitHub fine-grained PATs
	regexp.MustCompile(`(?i)^Bearer\s+\S+`),                 // Bearer tokens
	regexp.MustCompile(`^xox[bpsa]-[a-zA-Z0-9\-]+`),        // Slack tokens
	regexp.MustCompile(`^eyJ[a-zA-Z0-9_\-]+\.eyJ`),         // JWTs
	regexp.MustCompile(`^AKIA[A-Z0-9]{16}`),                 // AWS access keys
	regexp.MustCompile(`^npm_[a-zA-Z0-9]{36}`),              // npm tokens
}

const redacted = "[REDACTED]"

func isSensitiveKey(key string) bool {
	for _, p := range sensitiveKeyPatterns {
		if p.MatchString(key) {
			return true
		}
	}
	return false
}

func isSensitiveValue(value string) bool {
	for _, p := range sensitiveValuePatterns {
		if p.MatchString(value) {
			return true
		}
	}
	return false
}

// RedactSensitiveValues recursively traverses obj and replaces values that
// look like secrets (API keys, tokens, JWTs, bearer headers) with "[REDACTED]".
// Returns a new JsonObject — does not mutate the input.
func RedactSensitiveValues(obj JsonObject) JsonObject {
	result := make(JsonObject, len(obj))
	for key, value := range obj {
		// Special case: "env" key — redact all environment variable values wholesale.
		if key == "env" {
			if envMap, ok := value.(JsonObject); ok {
				redactedEnv := make(JsonObject, len(envMap))
				for envKey := range envMap {
					redactedEnv[envKey] = redacted
				}
				result[key] = redactedEnv
				continue
			}
		}

		// Sensitive key with a string value.
		if isSensitiveKey(key) {
			if strVal, ok := value.(string); ok {
				_ = strVal
				result[key] = redacted
				continue
			}
		}

		// Recurse into all values.
		result[key] = redactAny(value)
	}
	return result
}

// redactAny applies redaction logic to any value type.
func redactAny(value any) any {
	switch v := value.(type) {
	case string:
		if isSensitiveValue(v) {
			return redacted
		}
		return v
	case JsonObject:
		return RedactSensitiveValues(v)
	case []any:
		out := make([]any, len(v))
		for i, item := range v {
			out[i] = redactAny(item)
		}
		return out
	default:
		return value
	}
}
